# HG changeset patch
# User Betterbird <betterbird@betterbird.eu>
# Date 1736104618 -3600
# Parent  656b4e0c2d56583acfc7c65f8f451b424feaccd5
Bug 1890253 - Fix issue in nsMsgBrkMBoxStore::DiscardNewMessage().

Make DiscardNewMessage() truncate mbox to original size. Temporary fix.
Inspired by https://phabricator.services.mozilla.com/D213510 by Chiaki Ishikawa, adjusted by Betterbird.

diff --git a/mailnews/local/src/nsMsgBrkMBoxStore.cpp b/mailnews/local/src/nsMsgBrkMBoxStore.cpp
--- a/mailnews/local/src/nsMsgBrkMBoxStore.cpp
+++ b/mailnews/local/src/nsMsgBrkMBoxStore.cpp
@@ -41,8 +41,9 @@
 #include "mozilla/SlicedInputStream.h"
 #include "prprf.h"
 #include <cstdlib>  // for std::abs(int/long)
 #include <cmath>    // for std::abs(float/double)
+#include "nsIFile.h"
 
 /* for logging to Error Console */
 #include "nsIScriptError.h"
 
@@ -1284,10 +1285,24 @@ nsMsgBrkMBoxStore::DiscardNewMessage(nsI
     nsAutoCString folderURI;
     rv = folder->GetURI(folderURI);
     NS_ENSURE_SUCCESS(rv, rv);
     m_OutstandingStreams.remove(folderURI);
+  
+    // Truncate back to original size.
+    do {
+      nsAutoCString storeToken;
+      rv = aNewHdr->GetStoreToken(storeToken);
+      if (NS_FAILED(rv)) break;
+      uint64_t msgStart = storeToken.ToInteger64(&rv);
+      if (NS_FAILED(rv)) break;
+      nsCOMPtr<nsIFile> mboxFile;
+      rv = folder->GetFilePath(getter_AddRefs(mboxFile));
+      if (NS_FAILED(rv)) break;
+      rv = mboxFile->SetFileSize(msgStart);
+    } while (false);
   }
-  return NS_OK;
+
+  return rv;
 }
 
 NS_IMETHODIMP
 nsMsgBrkMBoxStore::FinishNewMessage(nsIOutputStream* aOutputStream,
